src/dev/GraphiQL.tsx

```diff
@@ -0,0 +1,218 @@
+      setConnectionMode("authenticated");
+    });
+
+    socket.on("reconnect", () => {
+      setConnectionMode("authenticating");
+      socket.emit("authenticate", { password: password });
+    });
+
+    socket.on("disconnect", () => {
+      setConnectionMode("disconnected");
+    });
+
+    if (socket.connected) {
+      authenticate();
+    }
+
+    return () => {
+      socket.off("connect");
+      socket.off("reconnecting");
+      socket.off("reconnect");
+      socket.off("reconnect_failed");
+      socket.off("disconnect");
+    };
+  }, [socket, password]);
+
+  if (connectionMode !== "authenticated") {
+    return <SplashScreen text={connectionMode} />;
+  }
+
+  return (
+    <RelayEnvironmentProvider environment={relayEnvironment}>
+      <AuthenticatedAppShellRenderer>{children}</AuthenticatedAppShellRenderer>
+    </RelayEnvironmentProvider>
+  );
+};
+
+const AuthenticatedGraphQLContent: React.FC<{
+  pcPassword: string;
+  localFetch: typeof fetch;
+}> = (props) => {
+  const socket = useSocket();
+  const ioGraphQLClient = createSocketIOGraphQLClient<any>(socket);
+  const ioFetcher = (graphQLParams: any) =>
+    ioGraphQLClient.execute({
+      ...graphQLParams,
+      operation: graphQLParams.query,
+    });
+
+  const fetcher = ioFetcher;
+  return (
+    <AuthenticatedGraphQLShell
+      password={props.pcPassword}
+      socket={socket}
+      role="DM"
+    >
+      <div style={{ height: "100vh" }}>
+        <DefaultGraphiQL fetcher={fetcher} />
+      </div>
+    </AuthenticatedGraphQLShell>
+  );
+};
+
+export const GraphiQL: React.FC<{
+  password: string | null;
+}> = (props) => {
+  const [pcPassword, setPcPassword] = usePcPassword();
+  const initialPcPassword = React.useRef(pcPassword);
+  let usedPassword = pcPassword;
+  // the password in the query parameters has priority.
+  if (pcPassword === initialPcPassword.current && props.password) {
+    usedPassword = props.password;
+  }
+
+  const [mode, setMode] = React.useState("LOADING");
+
+  const localFetch = React.useCallback(
+    (input, init = {}) => {
+      return fetch(buildApiUrl(input), {
+        ...init,
+        headers: {
+          Authorization: usedPassword ? `Bearer ${usedPassword}` : undefined,
+          ...init.headers,
+        },
+      }).then((res) => {
+        if (res.status === 401) {
+          console.error("Unauthenticated access.");
+          setMode("AUTHENTICATE");
+        }
+        return res;
+      });
+    },
+    [usedPassword]
+  );
+
+  useAsyncEffect(
+    function* () {
+      const result: any = yield localFetch("/auth").then((res) => res.json());
+      if (!result.data.role) {
+        setMode("AUTHENTICATE");
+        return;
+      }
+      setMode("READY");
+    },
+    [localFetch]
+  );
+
+  if (mode === "LOADING") {
+    return <SplashScreen text="Loading..." />;
+  }
+
+  if (mode === "AUTHENTICATE") {
+    return (
+      <AuthenticationScreen
+        requiredRole="PC"
```

Comment:
It should be possible to authenticate either as DM or PC. Maybe you can just add some kind of toggle?

Reply:
I'm trying to look around, but I can't figure out quite how I should implement a toggle. It seems like the user is automatically logged in based on whether they are on the /dm route but I haven't verified that.

Should I add a toggle in the UI that the user can click? If so, which screen/component should I add that on? in the player-area.tsx line 593 it also has a hardcoded "PC" set to requiredRole. (I know because they're a player, but yeah I might need a little elaboration if you don't mind)

is requiredRole={"PC" || "DM"}  something like what you're looking for?

Reply:
Yes it must be possible to use thr GraphQL api via the playground either as a player or dm, so permission related results can be tested. You could use the toggle component which we are already using (from chakra ui) and based on the true or false value either forward DM or PC